packages/ramps-controller/src/RampsController.ts

```diff
@@ -0,0 +1,233 @@
+    state,
+  }: {
+    messenger: RampsControllerMessenger;
+    state: Partial<RampsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: rampsControllerMetadata,
+      state: { ...defaultState, ...state } as RampsControllerState,
+    });
+
+    this.#registerMessageHandlers();
+  }
+
+  /**
+   * Gets the non-cached API URL based on the metamask environment.
+   *
+   * @param service - The API service to append to the base URL.
+   * @returns The non-cached API URL based on the metamask environment.
+   */
+  #getApiUrl(service?: ApiService): string {
+    let url = 'http://localhost:3000';
+    if (this.state.metamaskEnvironment === SdkEnvironment.Production) {
+      url = 'https://on-ramp.api.cx.metamask.io';
+    } else if (this.state.metamaskEnvironment === SdkEnvironment.Staging) {
+      url = 'https://on-ramp.uat-api.cx.metamask.io';
+    }
+
+    const urlWithPath = new URL(service ?? '', url);
+    return urlWithPath.toString();
+  }
+
+  async #getGeolocation(): Promise<string> {
+    try {
+      const url = this.#getApiUrl();
+      const response = await fetch(new URL('geolocation', url).toString());
+
+      if (!response.ok) {
+        throw new Error(`Failed to fetch geolocation: ${response.statusText}`);
+      }
+
+      const data = await response.json();
+      return data;
+    } catch (error) {
+      console.error('Error fetching geolocation:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Gets the countries for the given geolocation.
+   *
+   * @returns The countries for the given geolocation.
+   */
+  async getCountries(): Promise<void> {
+    try {
+      const geolocation = await this.#getGeolocation();
+      const url = this.#getApiUrl(ApiService.Regions);
+      const response = await fetch(
+        new URL(`countries/${geolocation}`, `${url}/`).toString(),
+      );
+
+      if (!response.ok) {
+        throw new Error(`Failed to fetch country data: ${response.statusText}`);
+      }
+
+      const data = await response.json();
+
+      // Validate expected data properties to handle invalid responses
+      if (
+        typeof data !== 'object' ||
+        data === null ||
+        typeof data.deposit !== 'boolean' ||
+        typeof data.aggregator !== 'boolean' ||
+        typeof data.global !== 'boolean'
+      ) {
+        throw new Error('Invalid country data format');
+      }
+
+      this.update((state) => {
+        state.region = {
+          id: geolocation as string,
+          deposit: data.deposit,
+          aggregator: data.aggregator,
+          global: data.global,
+        };
+      });
+    } catch (error) {
+      // Handle or re-throw as needed; here we simply log
+      console.error('Error in getCountries:', error);
+      // Optionally, clear region on error:
+      this.update((state) => {
+        state.region = null;
+      });
+      // Or rethrow if you want the caller to handle it:
+      // throw error;
```

Comment:
Generally in controllers and services we don't recommend swallowing errors like this unless there is a good reason to do so. Once again it would be helpful to understand how this would be used in practice.